src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCServletService.java

```diff
@@ -235,37 +245,61 @@ public void authenticateAndAuthorizeGitRequest(HttpServletRequest request, Repos
         log.debug("Authorizing user {} for repository {} took {}", user.getLogin(), localVCRepositoryUri, TimeLogUtil.formatDurationFrom(timeNanoStart));
     }
 
-    private User authenticateUser(String authorizationHeader) throws LocalVCAuthException {
+    private User authenticateUser(String authorizationHeader, ProgrammingExercise exercise, LocalVCRepositoryUri localVCRepositoryUri) throws LocalVCAuthException {
 
         UsernameAndPassword usernameAndPassword = extractUsernameAndPassword(authorizationHeader);
 
         String username = usernameAndPassword.username();
         String password = usernameAndPassword.password();
-
-        var user = userRepository.findOneByLogin(username);
+        User user = userRepository.findOneByLogin(username).orElseThrow(LocalVCAuthException::new);
 
         try {
             SecurityUtils.checkUsernameAndPasswordValidity(username, password);
+        }
+        catch (AccessForbiddenException | AuthenticationException e) {
+            throw new LocalVCAuthException();
+        }
 
-            // Note: we first check if the user has used a vcs access token instead of a password
+        // check user VCS access token
+        if (Objects.equals(user.getVcsAccessToken(), password) && user.getVcsAccessTokenExpiryDate() != null && user.getVcsAccessTokenExpiryDate().isAfter(ZonedDateTime.now())) {
+            return user;
+        }
 
-            if (user.isPresent() && !StringUtils.isEmpty(user.get().getVcsAccessToken()) && Objects.equals(user.get().getVcsAccessToken(), password)) {
-                // user is authenticated by using the correct access token
-                return user.get();
+        // Note: we first check if the user has used a vcs access token instead of a password
+        if (password.startsWith(TOKEN_PREFIX) && password.length() == VCS_ACCESS_TOKEN_LENGTH) {
```

Comment:
Question: Does this always work if I as a user define a password that matches these conditions?

Reply:
I tested it, and yes it works :+1: If the user's password is a token, the authentication using the password as a token fails at line 288. After that the password is checked, and the authentication succeeds.